feat(utils): Add loadable

Pinpickle opened the pull request:
Fixes: https://github.com/pmndrs/jotai/issues/672 (or at least tries to)

This implements loadable as discussed in the mentioned issue.

I can't get all edge-cases to work. In particular, recovering from error, when an atom goes through these states:

`suspense -> error -> suspense -> resolved`

Going from `error -> suspense` does not trigger a re-render.

I think this might need some changes to core, as I suspect by catching the error/promise, we are preventing some state-keeping logic from happening.

@dai-shi could you have a look? It's the last test in `loadable.test.tsx` that's failing.

dai-shi commented:
Nice work.
I have some coding preference to apply, reducing bundle size, fixing refAtom usage, but in general, it looks good.

However:

> I can't get all edge-cases to work. In particular, recovering from error

Oh, right... This is not resolved yet, if it's same as #601.
Yeah, I agree it's probably core responsibility.
I think it's going to take some time for this.
Should we fix it first, or could `loadable` be useful with this limitation for now?

Pinpickle commented:
@dai-shi I wasn't aware that it was a bug for all atoms

Given that this is a bug with all atoms, I _think_ it's OK to release as-is. It should probably be documented somewhere that error recovery doesn't work. Should I mark the test as skipped or remove altogether?

dai-shi commented:
I'm not 100% sure if it's the same bug as #601. (tbh, I haven't spent much time on it yet.)

Please leave the tests as skipped. Let's tackle it separately, if this is useful as is.
I'll request some changes on code.

dai-shi commented on src/utils/loadable.ts:
```diff
@@ -0,0 +1,67 @@
+
+const loadableAtomCache = new WeakMap()
+
+export type Loadable<Value> =
+  | LoadableLoading
+  | LoadableError
+  | LoadableValue<Value>
```
We prefer simpler (highly subjective though) type aliases, and avoid exporting types unless it's a huge pain. Let's see how it goes without exporting it.
```suggestion
type Loadable<Value> =
  | { state: 'loading' }
  | { state: 'hasError'; error: unknown }
  | { state: 'hasData'; data: Value }
```

Pinpickle replied:
Will do

dai-shi commented on src/utils/loadable.ts:
```diff
@@ -0,0 +1,67 @@
+  data: Value
+}
+
+const LOADING_LOADABLE: LoadableLoading = { state: 'loading' }
```
I'm guessing we don't need this if we have refAtom.

dai-shi commented on src/utils/loadable.ts:
```diff
@@ -0,0 +1,67 @@
+    return cachedAtom as Atom<Loadable<Value>>
+  }
+
+  const refAtom = atom<{ error: LoadableError | null }>({
+    error: null,
+  })
```
refAtom must be a read-only atom with a function for lazy initialization, otherwise the state could be shared among different providers.
I guess we can just store the entire `Loadable<Value>` type in the ref.
```suggestion
  const loadableRefAtom = atom((): [Loadable<Value>] => ([{ state: 'loading' }]))
```

Hmm, the tuple might be not very readable. In such a case, we can make it an object.

dai-shi replied:
Let me try writing the new derivedAtom with this ref.

```js
  const derivedAtom = atom((get): Loadable<Value> => {
    const loadableRef = get(loadableRefAtom)
    try {
      const value = get(anAtom)
      loadableRef[0] = { state: 'hasData', data: value }
    } catch (error) {
      if (error instanceof Promise) {
        if (loadableRef[0].state !== 'loading') {
          loadableRef[0] = { state: 'loading' }
        }
      } else {
        if (loadableRef[0].state !== 'hasError' || !Object.is(loadableRef[0].error, error)) {
          loadableRef[0] = { state: 'hasError', error }
        }
      }
    }
    return loadableRef[0]
  })
```

Now, I'm not sure if this is more readable, maybe your version is better.

---
Now, I have another idea. We can create a WeakMap at module level.
The key of the WeakMap is `error` object and the value is Loadable<Value> (= `{ state: 'hasError', error }`)
This allows us to eliminate the hacky refAtom.
Yeah, I think is far cleaner. What do you think?

Pinpickle replied:
I like it!

dai-shi reviewed (commented):
My final recommendation is to use WeakMap instead of refAtom, but let me shamelessly leave my intermediate comments.

Pinpickle commented:
@dai-shi I've addressed your comments, it's a lot cleaner now.

Let me know what you think.

I also looked into #601 and it is a core bug, opened a PR at #705

dai-shi reviewed (commented):
Looks very nice!
Added a small suggestion to refactor.

Pinpickle commented:
@dai-shi done! I did the same for the loadable cache as well, it's only one key

Pinpickle commented on tests/utils/loadable.test.tsx:
```diff
@@ -0,0 +1,150 @@
+  await findByText('Error: An error occurred')
+})
+
+it('loadable turns primitive throws into values', async () => {
```
As we're using a `WeakMap`, I added this test case. If a primitive value ever became a key for `error`, it wouldn't work.

dai-shi commented:
> I did the same for the loadable cache as well, it's only one key

Interestingly, I didn't notice it. Thanks. (There might be other utils we could do the same.)

dai-shi commented on src/utils/loadable.ts:
```diff
@@ -0,0 +1,52 @@
+import { atom } from 'jotai'
+import type { Atom } from 'jotai'
+
+const loadableAtomCache = new WeakMap<Atom<unknown>, Atom<Loadable<any>>>()
```
Can `any` be `unknown`? Or, TS complains?

Pinpickle replied:
It can be unknown, it'll just require a cast further down. I'll update

dai-shi reviewed (approved):
Looks good! Thanks so much.

dai-shi commented:
Would you also work on docs? For now, docs are in a separate repo for now.
https://github.com/pmndrs/website/tree/docs/docs/jotai/api/utils.mdx
(and, we should probably split utils page in to some pages.

Pinpickle commented:
@dai-shi yep I can have a look at that, will probably be next week

TwistedMinda commented:
@dai-shi Can't find any docs on loadable! It's available right?
I guess it's time to use the "Async" section of the docs to replace the good old "No Suspense".

Async section would look like:

- kick-off: atomWithPending (async then normal, as many times as you want)
  - quick example with suspense
  - quick example with loadable, mention that it can be used the same way for any async atom
  - quick mention (+ link) to the atomWithAsyncStorage that uses the same concept

- full async methods
   - async get example (use word "async selector" somehow so reader can relate)
   - async write example (use word "action" somehow so reader can relate)
   - async get-write example (i can't even relate to this)

- common practices
  - suspense-everywhere
  - preloading (mention it fits well with atomWithAsyncStorage)

What u think?
